### PR TITLE
Copy the content from the underlying image into the newly created volume.

### PR DESCRIPTION
**Pull request by vikas-goel**
The copy happens if prepare_volume_on_create is set to true in containers.conf file.

Docker does this by default.

Fixes: #10262

Signed-off-by: Vikas Goel <vikas.goel@gmail.com>

### CODE REVIEW

File: libpod/runtime_ctr.go
Comment (mheon) [inline libpod/runtime_ctr.go:464]: As discussed before, this *cannot* be the default. There will be significant performance implications from it.

Also, this must be following by a `ctr.cleanup()` to ensure we do not leak mountpoints and network configurations.

Comment (vikas-goel) [reply libpod/runtime_ctr.go:465]: I will add a `defer ctr.cleanup()` statement. After checking the `prepare()` code I thought calling it again will be a no-op whether it is invoked using `podman container init` or `podman run`.
For my knowledge, how will be create performance implications?

Comment (mheon) [reply libpod/runtime_ctr.go:476]: This adds another mount and unmount to the early-init code of Podman, plus a setup and teardown of the full CNI network stack. These are the two most expensive operations during container creation, and forcing the `podman run` command to do them twice is going to significantly increase runtime in typical cases.

The core issue here is that not all calls to container creation are `podman run`. It is perfectly legal for someone to `podman create` a container, wait a long time (years, even), and then `podman start` said container. We cannot keep the container's mounts and network namespace alive during this time - the end result of container creation must be a container with no resources consumed. This is why we must clean up after the initial mount of the container when doing this copy-up, and why it cannot be the default.

Comment (vikas-goel) [reply libpod/runtime_ctr.go:476]: > the end result of container creation must be a container with no resources consumed

Thank you @mheon . Agree with this.

File: libpod/runtime_ctr.go
Comment (cdoern) [inline libpod/runtime_ctr.go:472]: we should probably return this error instead of simply logging it.

**Review by cdoern — CHANGES_REQUESTED** [review]

File: libpod/runtime_ctr.go
Comment (vikas-goel) [reply libpod/runtime_ctr.go:472]: This is happening in a `defer` call where the dependency is on the function return error `retErr`. This piece of code handling is similar to how is handled elsewhere. I got this code from `libpod/container_internal.go`
```
 249 func (c *Container) handleRestartPolicy(ctx context.Context) (_ bool, retErr error) {
        ...
 277         defer func() {
 278                 if retErr != nil {
 279                         if err := c.cleanup(ctx); err != nil {
 280                                 logrus.Errorf("error cleaning up container %s: %v", c.ID(), err)
 281                         }
 282                 }
 283         }()
 284         if err := c.prepare(); err != nil {
 285                 return false, err
 286         }
 287
```

File: libpod/runtime_ctr.go
Comment (vikas-goel) [reply libpod/runtime_ctr.go:465]: WIP https://github.com/containers/common/pull/676

**Review by Luap99 — COMMENTED** [review]
I think we should split the network setup out from this, you only need storage. CNI setup is expensive. I also think in case of slirp4netns it will leak the process. cleanup() will not kill the slirp process.
Does `c.mountStorage()` and `c.cleanupStorage()` works for you?

**Comment by vikas-goel** [comment]
> I think we should split the network setup out from this, you only need storage. CNI setup is expensive. I also think in case of slirp4netns it will leak the process. cleanup() will not kill the slirp process.
> Does `c.mountStorage()` and `c.cleanupStorage()` works for you?

That will work. Thank you, @Luap99 . I had looked up that option earlier but thought there was some complexity. Not sure why. Checking again after your comment, it seems to be fine.

Currently, I am blocked and waiting for someone to merge the https://github.com/containers/common/pull/678 PR before I can make this change.

**Comment by vikas-goel** [comment]
/c/common

**Comment by vikas-goel** [comment]
Hi @mheon ,
| The updated c/common will have to be vendored into Podman to be used once this is merged.

I am not familiar with the command/steps. Could you help me on this one?

**Comment by mheon** [comment]
In the top-level directory for the Podman git repo, run:
`go get github.com/containers/common@$COMMITHASH` (where $COMMITHASH is the hash of the commit you want to vendor)

Then run `make vendor-in-container` to finalize the changes and commit the result.

Please note that you cannot have symlinks in the path of your working directory when you run these - has to be an absolute path or the commands will misbehave.

**Comment by vikas-goel** [comment]
Thank you @mheon . Looks like a new version of common is not needed per your instruction.
The build is failing for something in containers/buildah

**Comment by mheon** [comment]
I think it is asking for Common - from the tests:
```
libpod/runtime_ctr.go:466:24: r.config.Containers.PrepareVolumeOnCreate undefined (type "github.com/containers/common/pkg/config".ContainersConfig has no field or method PrepareVolumeOnCreate)
```

**Comment by vikas-goel** [comment]
> I think it is asking for Common - from the tests:
> 
> ```
> libpod/runtime_ctr.go:466:24: r.config.Containers.PrepareVolumeOnCreate undefined (type "github.com/containers/common/pkg/config".ContainersConfig has no field or method PrepareVolumeOnCreate)
> ```

Correct. That I comment I made was for vendor PR failure.

**Comment by vikas-goel** [comment]
Now that vendor c/common is done, could someone add `/ok-to-test` message or start a `/retest`? It requires a "trusted user" review.

**Comment by mheon** [comment]
I think you need to rebase on top of latest main and force-push - that will pick up the change and force CI to rerun

**Comment by vikas-goel** [comment]
@mheon , @rhatdan ,
Do we have a data volume container image in the pipeline to that I can use to write a test case?

Or, do you think creating a configuration file with `prepare_volume_on_create = true` and creating a container will be good enough?

**Comment by mheon** [comment]
Copy-up triggers on any volume, including volumes from `-v` - our general test is to use `-v` to bind-mount a host directory over a directory in the container that we know contains files.

**Pull request closed without merging**
